src/ucp/tag/offload/rndv.c

```diff
@@ -0,0 +1,303 @@
+    const ucp_ep_config_key_lane_t *ep_lane_cfg;
+    const ucp_tl_resource_desc_t *rsc;
+
+    ucs_assert(rpriv->lane < UCP_MAX_LANES);
+    ep_lane_cfg = &params->ep_config_key->lanes[rpriv->lane];
+    rsc         = &context->tl_rscs[ep_lane_cfg->rsc_index];
+
+    ucp_proto_default_query(params, attr);
+    ucs_string_buffer_appendf(&strb, UCT_TL_RESOURCE_DESC_FMT,
+                              UCT_TL_RESOURCE_DESC_ARG(&rsc->tl_rsc));
+}
+
+ucp_proto_t ucp_tag_rndv_offload_proto = {
+    .name     = "tag/rndv/offload",
+    .desc     = "rndv offload",
+    .flags    = 0,
+    .init     = ucp_tag_rndv_offload_proto_init,
+    .query    = ucp_tag_rndv_offload_proto_query,
+    .progress = {ucp_tag_rndv_offload_proto_progress},
+    .abort    = ucp_proto_rndv_rts_abort
+};
+
+static ucs_status_t
+ucp_tag_rndv_offload_sw_proto_init(const ucp_proto_init_params_t *init_params)
+{
+    ucp_worker_h worker                      = init_params->worker;
+    ucp_context_h context                    = worker->context;
+    ucp_ep_config_t *ep_config               = &worker->ep_config[
+                                                     init_params->ep_cfg_index];
+    ucp_proto_rndv_ctrl_priv_t *rpriv        = init_params->priv;
+    ucp_proto_rndv_ctrl_init_params_t params = {
+        .super.super         = *init_params,
+        .super.latency       = 0,
+        .super.overhead      = 40e-9,
+        .super.cfg_thresh    = context->config.ext.rndv_thresh,
+        .super.cfg_priority  = 60,
+        .super.max_length    = SIZE_MAX,
+        .super.min_iov       = 0,
+        .super.min_frag_offs = UCP_PROTO_COMMON_OFFSET_INVALID,
+        .super.max_frag_offs = ucs_offsetof(uct_iface_attr_t,
+                                            cap.tag.eager.max_short),
+        .super.max_iov_offs  = UCP_PROTO_COMMON_OFFSET_INVALID,
+        .super.hdr_size      = 0,
+        .super.send_op       = UCT_EP_OP_EAGER_SHORT,
+        .super.memtype_op    = UCT_EP_OP_LAST,
+        .super.flags         = UCP_PROTO_COMMON_INIT_FLAG_RESPONSE,
+        .remote_op_id        = UCP_OP_ID_RNDV_RECV,
+        .unpack_time         = UCS_LINEAR_FUNC_ZERO,
+        .perf_bias           = context->config.ext.rndv_perf_diff / 100.0,
+        .mem_info.type       = init_params->select_param->mem_type,
+        .mem_info.sys_dev    = init_params->select_param->sys_dev
+    };
+    ucp_lane_index_t lane;
+
+    if (!ucp_tag_rndv_check_op_id(init_params) ||
+        (init_params->select_param->dt_class != UCP_DATATYPE_CONTIG) ||
+        !ucp_ep_config_key_has_tag_lane(init_params->ep_config_key)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (ucp_tag_rndv_offload_supported(init_params, &params) &&
+        (worker->context->config.ext.tm_sw_rndv == UCS_NO)) {
+        /* Do not completely disable SW rndv, because it is used when hw offload
+         * protocol can't be used. */
+        params.super.min_length = ep_config->tag.offload.max_rndv_zcopy + 1;
```

Comment:
we prefer to be independent from ep_config and rely only on init_params->ep_config_key

Reply:
can you please elaborate?

Reply:
when new protocols take over, we plan to remote most of ep_config content